nova_vm/src/ecmascript/builtins/text_processing/string_objects/string_prototype.rs

```diff
@@ -718,16 +718,140 @@ impl StringPrototype {
+            if is_reg_exp {
+                // i. Let flags be ? Get(searchValue, "flags").
+                // ii. Perform ? RequireObjectCoercible(flags).
+                // iii. If ? ToString(flags) does not contain "g", throw a TypeError exception.
+                todo!();
+            }
+
+            // c. Let replacer be ? GetMethod(searchValue, %Symbol.replace%).
+            let symbol = WellKnownSymbolIndexes::Replace.into();
+            let replacer = get_method(agent, search_value, symbol)?;
+
+            // d. If replacer is not undefined, Return ? Call(replacer, searchValue, « O, replaceValue »).
+            if let Some(replacer) = replacer {
+                return call_function(
+                    agent,
+                    replacer,
+                    search_value,
+                    Some(ArgumentsList(&[o, replace_value])),
+                );
+            }
+        }
+
+        // 3. Let s be ? ToString(O).
+        let s = to_string(agent, o)?;
+
+        // 4. Let searchString be ? ToString(searchValue).
+        let search_string = to_string(agent, search_value)?;
+
+        // 5. Let functionalReplace be IsCallable(replaceValue).
+        if let Some(functional_replace) = is_callable(replace_value) {
+            // 7. Let searchLength be the length of searchString.
+            let search_length = search_string.len(agent);
+
+            // 8. Let advanceBy be max(1, searchLength).
+            let advance_by = max(1, search_length);
+
+            // 9. Let matchPositions be a new empty List.
+            let mut match_positions: Vec<usize> = vec![];
+
+            // 10. Let position be StringIndexOf(s, searchString, 0).
+            let search_str = search_string.as_str(agent).to_owned();
+            let subject = s.as_str(agent).to_owned();
+
+            let mut position = if let Some(position) = subject.find(&search_str) {
+                position
+            } else {
+                // If position is not-found, return s.
+                return Ok(s.into_value());
+            };
+
+            match_positions.push(position);
+            position += advance_by;
+
+            // 11. Repeat, while position is not not-found,
+            loop {
+                if let Some(pos) = subject[position..].find(&search_str) {
+                    match_positions.push(position + pos);
+                    position += advance_by;
+                } else {
+                    break;
+                };
+            }
+
+            // 12. Let endOfLastMatch be 0.
+            let mut end_of_last_match = 0;
+
+            // 13. Let result be the empty String.
+            let mut result = "".to_string();
+
+            // 14. For each element p of matchPositions, do
+            for p in match_positions {
+                // a. Let preserved be the substring of string from endOfLastMatch to p.
+                let preserved = subject[end_of_last_match..p].to_owned();
```

Comment:
**issue:** We can move this after the functional_replace call without the change being observable by callers, I believe. That allows us to also avoid the `to_owned` call.